### ISSUE TITLE
how to extract verb phrases for a given children?

### ISSUE THREAD

**Issue by randomgambit** [vanatteveldt/rsyntax#11]
Hello @vanatteveldt 

I hope all is well! Its been a while since the first version of this great package! I was trying a few things and I was not sure about how to extract the corresponding VERB + ADV or VERB + ADJ for a given subject.

See below an example for `the pressure has remained low for some time`. I would like to extract `remained low` but I was unable to do so with `rsyntax`. 

After `tquery`, remained low are not even in the same `fill` group...

```
example <- spacy_parse('the pressure has remained low for some time.', dependency = TRUE,
            tag = TRUE) %>% 
  as_tokenindex() 

> tquery(pos = 'VERB', label = 'hd',
+        children(lemma = 'pressure')) %>% 
+   annotate_tqueries(example,'ann',query_name = .)
   doc_id sentence token_id    token    lemma   pos tag parent relation entity
1:  text1        1        1      the      the   DET  DT      2      det       
2:  text1        1        2 pressure pressure  NOUN  NN      4    nsubj       
3:  text1        1        3      has     have   AUX VBZ      4      aux       
4:  text1        1        4 remained   remain  VERB VBN     NA     ROOT       
5:  text1        1        5      low      low   ADJ  JJ      4    acomp       
6:  text1        1        6      for      for   ADP  IN      4     prep       
7:  text1        1        7     some     some   DET  DT      8      det       
8:  text1        1        8     time     time  NOUN  NN      6     pobj       
9:  text1        1        9        .        . PUNCT   .      4    punct       
   ann               ann_id ann_fill
1:  hd query_name#text1.1.4        2
2:  hd query_name#text1.1.4        1
3:  hd query_name#text1.1.4        1
4:  hd query_name#text1.1.4        0
5:  hd query_name#text1.1.4        1
6:  hd query_name#text1.1.4        1
7:  hd query_name#text1.1.4        3
8:  hd query_name#text1.1.4        2
9:  hd query_name#text1.1.4        1
```

Do you have any ideas?
Thanks!

**Comment by kasperwelbers** [comment]
Hi @randomgambit,

In the current query, you are looking for a VERB that has a child with the specific lemma "pressure", and you only label the VERB. As a result, any node under this verb (in this case the entire sentence) is 'filled' with this label. Note that "remained" and "low" do in fact have the same label. The ann_fill column does not refer to fill groups, but to fill level (i.e. distance from the 'VERB').

This is easier to see with the plot_tree function:

```
example <- spacy_parse('the pressure has remained low for some time.', dependency = TRUE,
                       tag = TRUE) %>% 
  as_tokenindex() 

tq = tquery(pos = 'VERB', label = 'hd',
            children(lemma = 'pressure'))

annotate_tqueries(example,'ann', tq=tq) %>%
  plot_tree(annotation='ann')
```

I assume that what you are trying to do, is to specify that the verb label should be extended to ADJ of the verb. For this you could use more specific behavior for the fill heuristic. 

```
tq = tquery(pos = 'VERB', label = 'hd',
            fill(pos = 'ADJ', connected=T),
            children(lemma = 'pressure'))

annotate_tqueries(example,'ann', tq=tq) %>%
  plot_tree(annotation='ann')
```

For reference, the more intended approach here is to create some custom fill behaviors that you can use in different tqueries. For example, if you want to write some rules for extracting subject-verb relations, than you can first write fill heuristics for subjects and verbs. Here are some that work pretty well.

```
subject_fill = rsyntax::fill(NOT(relation=c('case','advmod','relcl','acl','acl:relcl','punct')), connected=T)
verb_fill = rsyntax::fill(relation=c('prt','aux','neg','punct','advmod','acomp'), pos = c('PART','VERB','AUX','ADJ','ADV'), connected=T)
```

Then you can plug these into the tquery

```
tq = tquery(label='verb', pos = 'VERB', verb_fill,
            children(label='subject', relation='nsubj', subject_fill))

annotate_tqueries(example,'ann', tq=tq) %>%
  plot_tree(annotation='ann')
```

Alternatively, you could also have given adjectives their own label.

```
tq = tquery(label='verb', pos = 'VERB', verb_fill,
            children(label='verb_adj', pos='ADJ'),
            children(label='subject', relation='nsubj', subject_fill))

annotate_tqueries(example,'ann', tq=tq) %>%
  plot_tree(annotation='ann')
```

So yeah, these queries tend to be complicated. The idea of separating the fill logic from the query is one of the more recent developments. We think it makes sense because you will often need multiple queries for a certain semantic relation (e.g., you need a separate pattern for subject-verb relations in a passive sentence), but the fill logic should often be the same. Anyway, Hope this helps.

**Comment by randomgambit** [comment]
Hi @kasperwelbers this is really great. 

However, I am running into trouble when using `fill` with the options you mentioned. Essentially, I am trying to create a way with `rstytax` to extract `noun-phrases` (https://stackoverflow.com/questions/33289820/noun-phrases-with-spacy) and `verb-phrases` (https://stackoverflow.com/questions/47856247/extract-verb-phrases-using-spacy)

For instance, extracting `the_great_vanatteveldt`, `10 year` and `this package` (noun-phrases) as well as `has spent` (verb phrases) in the example below

```
library(rsyntax)
library(tidyverse)
library(spacyr)
library(devtools)
#install_github("vanatteveldt/rsyntax")

spacy_initialize(model = 'en_core_web_lg')

mytext <- 'the great vanatteveldt has spent 10 years writing this package'

myparse <- spacy_parse(mytext, nounphrase = TRUE) %>% nounphrase_consolidate()

> myparse
  doc_id sentence_id token_id                  token                  lemma        pos
1  text1           1        1 the_great_vanatteveldt the_great_vanatteveldt nounphrase
2  text1           1        2                    has                   have        AUX
3  text1           1        3                  spent                  spend       VERB
4  text1           1        4               10_years                10_year nounphrase
5  text1           1        5                writing                  write       VERB
6  text1           1        6           this_package           this_package nounphrase
```

However using the `rsyntax` you proposed creates an error. 

```
# VERB PHRASES
> tq = tquery(pos = 'VERB', label = 'hd',
+             fill(pos = 'ADJ', connected=T))
Error: 2 components of `...` had unexpected names.

We detected these problematic arguments:
* `pos`
* `connected`

Did you misspecify an argument?
Run `rlang::last_error()` to see where the error occurred.
```

**Comment by kasperwelbers** [comment]
Hi @randomgambit,

Looks like a namespace problem. Probably there is a fill function somewhere in the tidyverse. If so, you should be able to use rsyntax::fill().

**Comment by randomgambit** [comment]
you are totally right! thanks for catching that. I would mention this on the `github` page given how ubiquitous `dplyr::fill` is...
I just have a clarification questions if you have a few moments.

In the example below I am able to pull the `verb phrase` "_has been increasing dramatically_" using `tquery`. 

What I do not understand actually is the meaning of `ann_fill`. Can you explain in words what it does? Are we getting a value of zero for all the immediate **children** of the `pos` we look for in the `tquery`? 

If so, what is the advantage of specifying the `rsyntax::fill()` exceptions that you proposed above?

Thanks!
```

mytext <- 'Over the years, the difficulty has been increasing dramatically.'

myparse <- spacy_parse(mytext, nounphrase = TRUE, pos = TRUE,dependency = TRUE)

tq = tquery(pos = 'VERB', label = 'hd',
            children(OR(pos = 'AUX', pos = 'ADV'), label = 'r'))

annotate_tqueries(myparse,'ann', tq=tq) 
    doc_id sentence token_id        token        lemma   pos parent relation entity nounphrase
 1:  text1        1        1         Over         over   ADP      9     prep                  
 2:  text1        1        2          the          the   DET      3      det DATE_B        beg
 3:  text1        1        3        years         year  NOUN      1     pobj DATE_I   end_root
 4:  text1        1        4            ,            , PUNCT      9    punct                  
 5:  text1        1        5          the          the   DET      6      det               beg
 6:  text1        1        6   difficulty   difficulty  NOUN      9    nsubj          end_root
 7:  text1        1        7          has         have   AUX      9      aux                  
 8:  text1        1        8         been           be   AUX      9      aux                  
 9:  text1        1        9   increasing     increase  VERB     NA     ROOT                  
10:  text1        1       10 dramatically dramatically   ADV      9   advmod                  
11:  text1        1       11            .            . PUNCT      9    punct                  
    whitespace ann       ann_id ann_fill
 1:       TRUE  hd tq#text1.1.9        1
 2:       TRUE  hd tq#text1.1.9        3
 3:      FALSE  hd tq#text1.1.9        2
 4:       TRUE  hd tq#text1.1.9        1
 5:       TRUE  hd tq#text1.1.9        2
 6:       TRUE  hd tq#text1.1.9        1
 7:       TRUE   r tq#text1.1.9        0
 8:       TRUE   r tq#text1.1.9        0
 9:       TRUE  hd tq#text1.1.9        0
10:      FALSE   r tq#text1.1.9        0
11:      FALSE  hd tq#text1.1.9        1
```

**Comment by kasperwelbers** [comment]
So, the idea of fill is that it recursively labels all children of a node with the same label. So, it first look for children (that meet the conditions specified in fill), then the children of these children, etc. ann_fill tells you whether the ann level was a direct match (0), child (1), grandchild (2), etc. 

So, in the example the token "increasing" has ann_fill 0 because it was a direct match for the 'hd' label. For "has", "been" and "dramatically" it's also 0, because it's a direct match for the 'r' label. But note that the verb and the AUX/ADJ now do not have the same label. 

The difference between using children or fill in this case, is whether you want to give the AUX and ADV the same label as the verb, or use a different label. Also, in the current query, you will only find cases where the verb actually has a AUX or ADV child. So if the verb does not have a AUX or ADV (e.g., Bob said hi), it will not match. With fill, the AUX or ADV is optional. Whether you should use children() or fill() thus depends on what you're trying to do. 

Generally speaking, you should think of fill as a way to deal with semantic role labels that can (but do not necessarily) cover multiple nodes. For instance, names often have compounds ("Donald Trump"). If you match one ("Trump"), you also want to include the compound "Donald" in the same label. For verbs, you might want to include ADJ and AUX so that "has been increasing dramatically" is grouped together. Then you would use fill(). Alternatively, you might want to specifically look for verbs with an adverb (ignoring verbs without an adverb), and give the adverb a different label. Then you would use children().

For sake of completeness, you can be more flexible, but off course that does make it more complicated. You can also say that children are optional by setting the req (required) argument to FALSE. You can also let children look recursively (children -> grandchildren -> etc) by increasing the depth.  

By the way, you can pass a vector of lookup values in a tquery (or children) function, which is more efficient that looking twice with or. So use pos = c('AUX','ADV') instead of OR(pos = 'aux', pos = 'ADV')

**Comment by randomgambit** [comment]
great. thanks @kasperwelbers. This is a great tool!!!

**Comment by randomgambit** [comment]
@kasperwelbers actually I think I found a problematic example (I hope you find this interesting and can help you improve the package). 

Consider this example where I am trying to parse separately `remain below trends` and `keep increasing`

```
mytext <- 'package creation is expected to remain below trends, but to keep increasing'

myparse <- spacy_parse(mytext, nounphrase = TRUE, pos = TRUE,dependency = TRUE)

tq <- tquery(pos = 'VERB',
             label = 'hrh', 
             rsyntax::fill(OR(pos = 'ADV', pos = 'ADJ', pos = 'AUX', pos = 'ADP', pos = 'PART', pos = 'VERB'), connected = TRUE),
             children(lemma = 'creation', label = 'h', OR( relation = 'nsubj',relation =  'nsubjpass')))

annotate_tqueries(myparse,'ann', tq=tq) 
    doc_id sentence token_id      token    lemma   pos parent  relation entity nounphrase whitespace  ann
 1:  text1        1        1    package  package  NOUN      2  compound               beg       TRUE    h
 2:  text1        1        2   creation creation  NOUN      4 nsubjpass          end_root       TRUE    h
 3:  text1        1        3         is       be   AUX      4   auxpass                         TRUE  hrh
 4:  text1        1        4   expected   expect  VERB     NA      ROOT                         TRUE  hrh
 5:  text1        1        5         to       to  PART      6       aux                         TRUE  hrh
 6:  text1        1        6     remain   remain  VERB      4     xcomp                         TRUE  hrh
 7:  text1        1        7      below    below   ADP      6      prep                         TRUE  hrh
 8:  text1        1        8     trends    trend  NOUN      7      pobj          beg_root      FALSE <NA>
 9:  text1        1        9          ,        , PUNCT      6     punct                         TRUE <NA>
10:  text1        1       10        but      but CCONJ      6        cc                         TRUE <NA>
11:  text1        1       11         to       to  PART     12       aux                         TRUE  hrh
12:  text1        1       12       keep     keep  VERB      6      conj                         TRUE  hrh
13:  text1        1       13 increasing increase  VERB     12     xcomp                        FALSE  hrh
          ann_id ann_fill
 1: tq#text1.1.4        1
 2: tq#text1.1.4        0
 3: tq#text1.1.4        1
 4: tq#text1.1.4        0
 5: tq#text1.1.4        2
 6: tq#text1.1.4        1
 7: tq#text1.1.4        2
 8:         <NA>       NA
 9:         <NA>       NA
10:         <NA>       NA
11: tq#text1.1.4        3
12: tq#text1.1.4        2
13: tq#text1.1.4        3
```

As you can see, there is nothing in the `rsyntax` output that allows me to easily extract the two verb phrases. Indeed, the annotation id is the same for both group, although the plot tree shows that both verb phrases are distinct. 

In other words, if I were to `group_by(annotation_id)` and `paste(collapse = '')` the lemmas I would obtain the string "**is expected to remain below to keep increasing**". 

Is there a `rsyntax` solution to this issue?
Thanks!!

![image](https://user-images.githubusercontent.com/8282510/96668490-46f54c80-1329-11eb-8b20-e5e86b42e103.png)

**Comment by kasperwelbers** [comment]
Hi @randomgambit 

This is indeed an interesting case, and one we are very aware of. 

More generally put, this is a nested sentence (a conjunction in this case), and a case of "argument drop". The problem is that some of the arguments are 'implicit'. Note that these are two sentences: "Package creation is expected to remain below trends" and "Package creation is expected to keep increasing". In the second one, the "package creation is expected" part is implied by convention. 

The original solution that Wouter used is to write some additional rules for the most common of these cases. For every tquery, you can make an additional tquery that "looks around the corner" by including the conjunction, so to say. However, this quickly requires many tqueries. Also, note that we cannot annotate the tokens data properly this way. We would have to label "package creation is expected" twice, once for "to remain below trends", and once for "to keep increasing".

The alternative we're working on is to first transform dependency trees to resolve argument drop. We then copy implicit parts of the tree and break it into multiple trees. Effectively, this is a form of text simplification, which is also sometimes done with dependency tree operations. rsyntax actually has functcionalities for tree transformations (see working paper on github for details).

However, this is where rsyntax gets even more complicated, so my current goal is to provide code to deal with the most common cases in a soon to come rsyntaxRecipes repository. I've put my current progress in a gist for you, so you can try it out, but note that this is work in progress:

[Spacy transformation rules](https://gist.github.com/kasperwelbers/93b7c465cd3fb8ce5805843cd6fca245)

You can run this code, and then apply spacy_simplify() on your tokens. 

Owh, by the way. In the future, please make separate issues for different questions. These issues help us think about how to advance rsyntax, but for documentation its nicer to have these discussions in separate threads.

**Comment by randomgambit** [comment]
Thanks! I will be sure to open other queries for other questions then. 

Yesterday, I came up with some "hack" that seems to work. The idea is simply to "relabel" the different `ann_id` groups manually (using the variable `myid`) that looks whether `ann_id` is following an empty `ann_id` (which means this is a different group) or not.

See below:

```
mytext <- 'package creation is expected to remain below trends, but to keep increasing'

myparse <- spacy_parse(mytext, nounphrase = TRUE, pos = TRUE,dependency = TRUE)

tq <- tquery(pos = 'VERB',
             label = 'hrh', 
             rsyntax::fill(OR(pos = 'ADV', pos = 'ADJ', pos = 'AUX', pos = 'ADP', pos = 'PART', pos = 'VERB'), connected = TRUE),
             children(lemma = 'creation', label = 'h', OR( relation = 'nsubj',relation =  'nsubjpass')))


annotate_tqueries(myparse,'ann', tq=tq) %>% 
  mutate(ann_id = as.character(ann_id),
         ann_id = replace_na(ann_id, 'zero'),
         ch = if_else(ann_id != dplyr::lag(ann_id), 1, 0),
         ch = replace_na(ch,0),
         myid = cumsum(ch)) %>% 
  dplyr::filter(ann == 'hrh') %>% 
  group_by(doc_id, sentence, myid) %>% summarise(extract = paste(token, collapse = ' '))
`summarise()` regrouping output by 'doc_id', 'sentence' (override with `.groups` argument)
# A tibble: 2 x 4
# Groups:   doc_id, sentence [1]
  doc_id sentence  myid extract                    
  <chr>     <int> <dbl> <chr>                      
1 text1         1     0 is expected to remain below
2 text1         1     2 to keep increasing    
```

What do you think? Would this work in general?

**Comment by kasperwelbers** [comment]
Sorry that this got stranded. I'll have another look at the problem when I hopefully get around to spending some more time on this package soon!
